### DISCUSSION TITLE
Se me dispara el error y el entrenamiento es muy lento

### DISCUSSION THREAD

**Discussion by Dunkan1** [JavierBZ/Tarea3-IIC3697#13] [Q&A] [answered]
![Screenshot from 2021-06-23 18-19-34](https://user-images.githubusercontent.com/26102365/123175398-bb036980-d44f-11eb-9a8c-6709dd9ca93f.png)

Hola, estaba probando el entrenamiento cuando me di cuenta de que tarda mucho mas de lo que deberia, segun lo que tenia entendido era de ~30 min por epoca pero a mi se me demora casi 2 horas, pienso que puede ser culpa de un pequeño for en el sample del ReplayMemory para entregar lo datos bien (que no retorne asi [(x,y),(x1,y1)] sino que asi [(x,x1),(y,y1)] ), pero no se si pueda ser otra cosa. Por otro lado la perdida se dispara, no se si al resto le pasa lo mismo (el unico paso que no he aplicado es normalizar)

**Comment by JavierBZ** [comment]
Hola, no entiendo bien tu ejemplo del sample, pero lo que debes hacer es retornar un arreglo en donde la primera dimensión es la de batch para cada elemento de la experiencia (por ej obst_t.shape -> (32,4,84,84), act_t.shape -> (32,), etc). Intenta que tu función sample no tenga ningun for, eso debe estar enlentenciendo bastante la ejecución.

Lo otro esa es tu primera epoca? porque si es así porque estas calculando loss en los primeros 50.000 pasos? recuerda que los primeros 50.000 pasos son sólo de exploración y no hay ningun calculo de loss.

La causa que se te dispare pueden ser varias, una de ellas la normalización, pero lo más común es que tengas algun error en el cálculo de la loss. Ten en cuenta todos los detalles de la fórmula, en especial que la loss se propaga sólo por la acción que llevo el estado obs a next_obs. Para asegurarte te sugiero que antes de entrenar primero compruebes tu calculo con tensores de prueba.

Espero te sirva!

**Comment by Pudindin** [comment, marked as answer]
Hola! Para evitar el for en el sample, puedes usar zip(), que hace justo lo que quieres hacer con el for.
De todas formas, sería bueno que midas el tiempo que tardan diversas partes del código, y así ver dónde está el problema. Mi entrenamiento tarda, en 1 época, entre 2 y 3 horas (y tuve que bajarle al tamaño del batch); y al imprimir los tiempos, noté que lo que más tarda es el optimizer.step() y el loss.backward(), no el buffer de memoria. He investigado, pero no he encontrado cómo mejorarlo (si alguien tiene un consejo, genial)

**Reply by Pudindin** [reply]
Update: resulta que no había configurado colab para usar GPU; estaba usando CPU y por eso tardaba tanto.

**Reply by Dunkan1** [reply]
ahora cuanto tarda el tuyo?, agregue el for con el zip() tal como indica aqui [https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html](url) y se demora 2:48:28 con 24.73it/s por epoca (tambien tiene la GPU activada)

**Reply by Pudindin** [reply]
A mí me va sobre las 100 it/s. ¿probaste imprimiendo los tiempos de ejecución de distintas partes del código? Para identificar qué se está demorando más

**Reply by Dunkan1** [reply]
eso voy a hacer a continuacion, gracias

**Reply by jdiazram** [reply]
Lograste bajar el tiempo por epoca?

**Reply by Dunkan1** [reply]
no, aun se demora, revise los tiempos de ejecucion y en lo que se demora mas (0.33 seg. por iteracion) es en la parte 2. Perform experience replay and train the network. de esa parte lo que mas se demora (0.26 seg. por iteracion) es en el sampleo y pasar a tensores. el sampleo lo tengo tal como indica en el link de pytorch
